src/Std/Data/DHashMap/Raw.lean

```diff
@@ -0,0 +1,438 @@
+/-
+Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Markus Himmel
+-/
+prelude
+import Init.Data.BEq
+import Init.Data.Hashable
+import Std.Data.DHashMap.Internal.Defs
+
+/-
+# Dependent hash maps with unbundled well-formedness invariant
+
+This file develops the type `Std.Data.DHashMap.Raw` of dependent hash
+maps with unbundled well-formedness invariant.
+
+This version is safe to use in nested inductive types. The well-formedness predicate is
+available as `Std.Data.DHashMap.Raw.WF` and we prove in this file that all operations preserve
+well-formedness. When in doubt, prefer `DHashMap` over `DHashMap.Raw`.
+
+Lemmas about the operations on `Std.Data.DHashMap.Raw` are available in the module
+`Std.Data.DHashMap.RawLemmas`.
+-/
+
+set_option linter.missingDocs true
+set_option autoImplicit false
+
+universe u v w
+
+variable {α : Type u} {β : α → Type v} {δ : Type w} {m : Type w → Type w} [Monad m]
+
+namespace Std
+
+open DHashMap.Internal DHashMap.Internal.List
+
+namespace DHashMap
+
+namespace Raw
+
+/-!
+The type `DHashMap.Raw` itself is defined in the module `Std.Data.DHashmap.RawDef` for import
+structure reasons.
+-/
+
+/-- Creates a new empty hash map. The optional parameter `capacity` can be supplied to presize the
+map so that it can hold the given number of mappings without reallocating. It is also possible to
+use the empty collection notations `∅` and `{}` to create an empty hash map with the default
+capacity. -/
+@[inline] def empty (capacity := 8) : Raw α β :=
+  (Raw₀.empty capacity).1
+
+instance : EmptyCollection (Raw α β) where
+  emptyCollection := empty
+
+/-- Insert the given mapping into the map, replacing an existing mapping for the key if there is
+one. -/
+@[inline] def insert [BEq α] [Hashable α] (m : Raw α β) (a : α) (b : β a) : Raw α β :=
+  if h : 0 < m.buckets.size then
+    (Raw₀.insert ⟨m, h⟩ a b).1
+  else m -- will never happen for well-formed inputs
+
+/-- If there is no mapping for the given key, insert the given mapping into the map. Otherwise,
+return the map unaltered. -/
+@[inline] def insertIfNew [BEq α] [Hashable α] (m : Raw α β) (a : α) (b : β a) : Raw α β :=
+  if h : 0 < m.buckets.size then
+    (Raw₀.insertIfNew ⟨m, h⟩ a b).1
+  else m -- will never happen for well-formed inputs
+
+/-- Equivalent to (but potentially faster than) calling `contains` followed by `insert`. -/
+@[inline] def containsThenInsert [BEq α] [Hashable α] (m : Raw α β) (a : α) (b : β a) :
+    Bool × Raw α β:=
+  if h : 0 < m.buckets.size then
+    let ⟨replaced, ⟨r, _⟩⟩ := Raw₀.containsThenInsert ⟨m, h⟩ a b
+    ⟨replaced, r⟩
+  else (false, m) -- will never happen for well-formed inputs
+
+/-- Equivalent to (but potentially faster than) calling `get?` followed by `insertIfNew`. -/
```

Comment:
```suggestion
/--
Checks whether a key is present in a map, and inserts a value for the key if not.

Equivalent to (but potentially faster than) calling `get?` followed by `insertIfNew`.
-/
```

Reply:
Did you mean `associate` or `associated`? My non-native english is failing me here :)

Reply:
My native English is what's failing. It should be "associated".